uberduck_ml_dev/exec/train_radtts_with_ray.py

```diff
@@ -0,0 +1,1480 @@
+        output_indices[indices] = new_indices
+    return output_indices
+
+
+def ray_df_preprocessing(df):
+    transcripts = df.transcript.tolist()
+    audio_bytes_list = df.audio_bytes.tolist()
+    speaker_ids = df.speaker_id.tolist()
+    paths = df.path.tolist()
+    f0_paths = df.f0_path.tolist()
+    audio_embeddings = df.audio_embedding.tolist()
+    # NOTE (Sam): I'm great at naming things.
+    # shuffle_indices = np.load('/usr/src/app/radtts/30shuffle_sdfixed_indices.pt.npy')[:100]
+    # shuffle_indices = get_shuffle_indices(speaker_ids)
+    # np.save('/usr/src/app/radtts/asdfasdfasdfasdfasdf.pt', shuffle_indices)
+    # audio_embeddings = [audio_embeddings[i] for i in shuffle_indices]
+    # audio_embeddings = audio_embeddings[:100]
+    collate_input = []
+    for transcript, audio_bytes, speaker_id, f0_path, audio_embedding in zip(
+        transcripts, audio_bytes_list, speaker_ids, f0_paths, audio_embeddings
+    ):
+        # print(datetime.now(), 'start')
+        # Audio
+        # print(datetime.now(), 'pre wav read and norm')
+        bio = BytesIO(audio_bytes)
+        sr, wav_data = wavfile.read(bio)
+        audio = torch.FloatTensor(wav_data)
+        # TODO (Sam): fix this for anyscale training
+        audio_norm = audio / (np.abs(audio).max() * 2)
+        # print(datetime.now(), 'pre text embed')
+        text_sequence = get_text(transcript)
+        # print(datetime.now(), 'pre mel compute')
+        mel = get_mel(audio_norm, data_config["max_wav_value"], stft)
+        mel = torch.squeeze(mel, 0)
+        # print(datetime.now(), 'pre f0 load')
+        dikt = torch.load(f0_path)
+        f0 = dikt["f0"]
+        p_voiced = dikt["p_voiced"]
+        voiced_mask = dikt["voiced_mask"]
+        # f0, voiced_mask, p_voiced = get_f0_pvoiced(
+        #     audio.cpu().numpy(), f0_min = data_config['f0_min'], f0_max=data_config["f0_max"], hop_length=data_config['hop_length'], frame_length=data_config['filter_length'], sampling_rate=22050)
+        f0 = f0_normalize(f0, f0_min=data_config["f0_min"])
+        # print(datetime.now(), 'pre energy compute')
+        energy_avg = get_energy_average(mel)
+        # print(datetime.now(), 'pre prior load')
+        prior_path = "{}_{}".format(text_sequence.shape[0], mel.shape[1])
+        prior_path = os.path.join("/usr/src/app/radtts/data_cache", prior_path)
+        prior_path += "_prior.pth"
+        attn_prior = torch.load(prior_path)
+        # attn_prior = get_attention_prior(text_sequence.shape[0], mel.shape[1])
+        speaker_id = get_speaker_id(speaker_id)
+        # datum = torch.FloatTensor(audio_norm).unsqueeze(-1).t().cuda()
+        # audio_embedding = torch.load(emb_path)
+        audio_embedding = torch.FloatTensor(audio_embedding)
+        # audio_embeddings = audio_encoder(datum)
+        # audio_embeddings = None
+        # NOTE (Sam): might be faster to return dictionary arrays of batched inputs instead of list
```

Comment:
started to try this - got to dictionary of dictionaries i think (ray doesnt have built in padding for time series) but didnt evaluate speed.